nixos/modules/services/misc/mautrix-telegram.nix

```diff
@@ -0,0 +1,62 @@
+{ lib, config, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.mautrix-telegram;
+
+  configFile = pkgs.runCommand "mautrix-telegram" {
+    buildInputs = [ pkgs.mautrix-telegram pkgs.remarshal ];
+    preferLocalBuild = true;
+  } ''
+    mkdir -p $out
+    remarshal -if json -of yaml \
+      < ${pkgs.writeText "config.json" (builtins.toJSON cfg.configOptions)} \
+      > $out/config.yaml
+
+    ${pkgs.mautrix-telegram}/bin/mautrix-telegram -c $out/config.yaml -g -r $out/registration.yaml
+  '';
+
+in
+{
+  options.services.mautrix-telegram = {
+    enable = mkEnableOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to enable mautrix-telegram, a puppeting bridge between Matrix and WhatsApp.
+
+        See <link xlink:href="https://github.com/tulir/mautrix-telegram/wiki/Bridge-setup">here</link> for documentation.
+     '';
+    };
+
+    configOptions = mkOption {
+      type = types.attrs;
+      description = ''
+        This options will be transform in YAML configuration file for the bridge
+
+        You can have the list of option <link xlink:href="https://github.com/tulir/mautrix-telegram/blob/master/example-config.yaml">here</link>
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.mautrix-telegram = {
+      description = "Mautrix-Telegram Service - A Telegram bridge for Matrix";
+      after = [ "network.target" "matrix-synapse.service" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        DynamicUser = true;
+        StateDirectory = "mautrix-telegram";
+        ExecStart = ''
+          ${pkgs.mautrix-telegram}/bin/mautrix-telegram -c "${configFile}/config.yaml"
```

Comment:
I'm afraid that this won't work right away. As described in [the upstream docs](https://github.com/tulir/mautrix-telegram/wiki/Bridge-setup#production-setup) you need to configure a database first using `alembic`.

To get this working locally, I had to patch the `mautrix-telegram` package accordingly: https://github.com/NixOS/nixpkgs/pull/59519

Reply:
I had an error locally to generate the database. I will try with your pull request.

Reply:
I try to merge your pull request but I've a problem during the generation of the database : 
```
Traceback (most recent call last):
  File "/nix/store/2z280p610w8yn9ypsngk2xzpgi54hl6b-python3.7-alembic-1.0.7/bin/.alembic-wrapped", line 11, in <module>
    sys.exit(main())
  File "/nix/store/2z280p610w8yn9ypsngk2xzpgi54hl6b-python3.7-alembic-1.0.7/lib/python3.7/site-packages/alembic/config.py", line 527, in main                                                      
    CommandLine(prog=prog).main(argv=argv)
  File "/nix/store/2z280p610w8yn9ypsngk2xzpgi54hl6b-python3.7-alembic-1.0.7/lib/python3.7/site-packages/alembic/config.py", line 521, in main                                                      
    self.run_cmd(cfg, options)
  File "/nix/store/2z280p610w8yn9ypsngk2xzpgi54hl6b-python3.7-alembic-1.0.7/lib/python3.7/site-packages/alembic/config.py", line 501, in run_cmd                                                   
    **dict((k, getattr(options, k, None)) for k in kwarg)
  File "/nix/store/2z280p610w8yn9ypsngk2xzpgi54hl6b-python3.7-alembic-1.0.7/lib/python3.7/site-packages/alembic/command.py", line 276, in upgrade                                                  
    script.run_env()
  File "/nix/store/2z280p610w8yn9ypsngk2xzpgi54hl6b-python3.7-alembic-1.0.7/lib/python3.7/site-packages/alembic/script/base.py", line 475, in run_env                                              
    util.load_python_file(self.dir, "env.py")
  File "/nix/store/2z280p610w8yn9ypsngk2xzpgi54hl6b-python3.7-alembic-1.0.7/lib/python3.7/site-packages/alembic/util/pyfiles.py", line 90, in load_python_file                                     
    module = load_module_py(module_id, path)
  File "/nix/store/2z280p610w8yn9ypsngk2xzpgi54hl6b-python3.7-alembic-1.0.7/lib/python3.7/site-packages/alembic/util/compat.py", line 156, in load_module_py                                       
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "alembic/env.py", line 10, in <module>
    from mautrix_telegram.base import Base
ModuleNotFoundError: No module named 'mautrix_telegram.base'
```

Reply:
Have you used `mautrix-telegram.alembic` or `python3Packages.alembic`? As described in the PR you need a `alembic` with `mautrix-telegram` available which is why I added the `passthru` expression.

I just checked it locally whether it creates a sqlite db successfully in my `mautrix-telegram` checkout.

Reply:
Something like this should work
```
preStart = ''
  ${pkgs.mautrix-telegram.alembic}/bin/alembic -x config=${configFile}/config.yaml upgrade head
'';
```

Reply:
@Ma27 Is it possible that even if I merge your commit in #59519, I continue to have this errors on generation of files ?

Reply:
What kind of errors? And how does your local state look like ATM?

Reply:
mai 04 00:38:50 SENNAS01 mautrix-telegram[27892]: Traceback (most recent call last):
mai 04 00:38:50 SENNAS01 mautrix-telegram[27892]:   File "/nix/store/3b610fcg302df6bj2hlrfmzxnrb5sd5j-python3.7-mautrix-telegram-0.5.0/lib/python3.7/site-packages/mautrix_telegram/__main__.py", line 138, in <module>
mai 04 00:38:50 SENNAS01 mautrix-telegram[27892]:     loop.run_until_complete(asyncio.gather(*startup_actions, loop=loop))                                                                         
mai 04 00:38:50 SENNAS01 mautrix-telegram[27892]:   File "/nix/store/0n8slcq8p5x31kc9hncabsqq9y3fpkzp-python3-3.7.3/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete         
mai 04 00:38:50 SENNAS01 mautrix-telegram[27892]:     return future.result()
mai 04 00:38:50 SENNAS01 mautrix-telegram[27892]:   File "/nix/store/3b610fcg302df6bj2hlrfmzxnrb5sd5j-python3.7-mautrix-telegram-0.5.0/lib/python3.7/site-packages/mautrix_telegram/abstract_user.py", line 178, in ensure_started
mai 04 00:38:50 SENNAS01 mautrix-telegram[27892]:     await self.start(delete_unless_authenticated=not even_if_no_session)                                                                         
mai 04 00:38:50  mautrix-telegram[27892]:   File "/nix/store/3b610fcg302df6bj2hlrfmzxnrb5sd5j-python3.7-mautrix-telegram-0.5.0/lib/python3.7/site-packages/mautrix_telegram/user.py", line 168, in start
mai 04 00:38:50  mautrix-telegram[27892]:     await super().start()
mai 04 00:38:50  mautrix-telegram[27892]:   File "/nix/store/3b610fcg302df6bj2hlrfmzxnrb5sd5j-python3.7-mautrix-telegram-0.5.0/lib/python3.7/site-packages/mautrix_telegram/abstract_user.py", line 168, in start
mai 04 00:38:50  mautrix-telegram[27892]:     self._init_client()
mai 04 00:38:50  mautrix-telegram[27892]:   File "/nix/store/3b610fcg302df6bj2hlrfmzxnrb5sd5j-python3.7-mautrix-telegram-0.5.0/lib/python3.7/site-packages/mautrix_telegram/abstract_user.py", line 118, in _init_client
mai 04 00:38:50  mautrix-telegram[27892]:     proxy=self._proxy_settings)
mai 04 00:38:50  mautrix-telegram[27892]:   File "/nix/store/dmgd30jav2fazayjvc3gzzqzmfawnbjn-python3.7-telethon-1.6.2/lib/python3.7/site-packages/telethon/client/telegrambaseclient.py", line 305, in __init__
mai 04 00:38:50 SENNAS01 mautrix-telegram[27892]:     self._state = self.session.get_update_state(0)                                                                                               
mai 04 00:38:50 SENNAS01 mautrix-telegram[27892]:   File "/nix/store/dpwmd64gvs487ipq27gph4d30frgplzi-python3.7-telethon-session-sqlalchemy-0.2.14/lib/python3.7/site-packages/alchemysession/core.py", line 41, in get_update_state
mai 04 00:38:50  mautrix-telegram[27892]:     date = datetime.datetime.utcfromtimestamp(date)                                                                                              
mai 04 00:38:50  mautrix-telegram[27892]: TypeError: an integer is required (got type str)
mai 04 00:38:50  systemd[1]: mautrix-telegram.service: Main process exited, code=exited, st


# ll /var/lib/mautrix-telegram/

total 799K
drwxr-xr-x 3 65043 65043    5 24 avril 09:28 .
drwx------ 5 root  root     5 22 avril 22:00 ..
drwxr-xr-x 4 65043 65043    5 11 févr. 21:18 alembic
-rw-r--r-- 1 65043 65043 1,6K 11 févr. 21:15 alembic.ini
-rw-r--r-- 1 65043 65043 2,7M 24 avril 09:27 mautrix-telegram.db

Reply:
With local state I actually meant what's currently in your local `nixpkgs` checkout. And can you please share the contents of your config file, please?

Also, please use [markdown](https://guides.github.com/features/mastering-markdown/) for code and log outputs, that makes it way easier to read IMHO.

Reply:
Here the default.nix file system : 
```
{ lib, python3, mautrix-telegram }:    
    
with python3.pkgs;    
    
buildPythonPackage rec {    
  pname = "mautrix-telegram";    
  version = "0.5.1";    
    
  src = fetchPypi {    
    inherit pname version;    
    sha256 = "51951845e52c4ca5410e0f4a51d99014dd6df2fcedfca8b7241e045359cbf112";    
  };    
        
  postPatch = ''    
    sed -i -e '/alembic>/d' setup.py    
  '';    
    
  propagatedBuildInputs = [    
    Mako    
    aiohttp    
    mautrix-appservice    
    sqlalchemy    
    CommonMark    
    ruamel_yaml    
    future-fstrings    
    python_magic    
    telethon    
    telethon-session-sqlalchemy    
    pillow    
    lxml    
  ];    
    
  # `alembic` (a database migration tool) is only needed for the initial setup,    
  # and not needed during the actual runtime. However `alembic` requires `mautrix-telegram`    
  # in its environment to create a database schema from all models.    
  #    
  # Hence we need to patch away `alembic` from `mautrix-telegram` and create an `alembic`    
  # which has `mautrix-telegram` in its environment.    
  passthru.alembic = alembic.overrideAttrs (old: {    
    propagatedBuildInputs = old.propagatedBuildInputs ++ [    
      mautrix-telegram    
    ];    
  });    
    
  checkInputs = [     pytest
    pytestrunner
    pytest-mock
    pytest-asyncio
  ];

  meta = with lib; {
    homepage = https://github.com/tulir/mautrix-telegram;
    description = "A Matrix-Telegram hybrid puppeting/relaybot bridge";
    license = licenses.agpl3Plus;
    maintainers = with maintainers; [ nyanloutre ma27 ];
  };
}

```

My config : 
```
    pytest
    pytestrunner
    pytest-mock
    pytest-asyncio
  ];

  meta = with lib; {
    homepage = https://github.com/tulir/mautrix-telegram;
    description = "A Matrix-Telegram hybrid puppeting/relaybot bridge";
    license = licenses.agpl3Plus;
    maintainers = with maintainers; [ nyanloutre ma27 ];
  };
}
    pytest
    pytestrunner
    pytest-mock
    pytest-asyncio
  ];

  meta = with lib; {
    homepage = https://github.com/tulir/mautrix-telegram;
    description = "A Matrix-Telegram hybrid puppeting/relaybot bridge";
    license = licenses.agpl3Plus;
    maintainers = with maintainers; [ nyanloutre ma27 ];
  };
}

```

Reply:
So, basically the change from my PR. I'm even more interested in how you generate the YAML config for mautrix-telegram (manually? If yes, *which* `alembic` is used?) or how the module currently looks like locally.

Reply:
I use a command to generate the database on startup

```
      serviceConfig = {    
        DynamicUser = true;    
        StateDirectory = "mautrix-telegram";    
        preStart = ''    
          ${pkgs.mautrix-telegram.alembic}/bin/alembic -x config=${configFile}/config.yaml upgrade head    
        '';    
        ExecStart = ''    
          ${pkgs.mautrix-telegram}/bin/mautrix-telegram -c "${configFile}/config.yaml"    
        '';    
        Restart = "on-failure";    
      };

```

Reply:
It's the module that I publish on my PR which is used locally :)

Reply:
yeah, I know that :)
Unfortunately I don't have the time ATM to find out what's going on here. Unless you're faster, I'd have a look at this next weekend :)

Reply:
Hello @Ma27, I went forward on this and found that alembic need a alembic folder present in 
> /nix/store/8pk3rga7dzjkpwwd17r94ya3v8pidggr-python3.7-mautrix-telegram-0.5.1/alembic

but I don't known how to use it because I don't where alembic is execute, and I can't surcharge the option in the ini file ([alembic] script_location)

Reply:
Have you tried setting the working directory to the application's source root? It might work as `script_location` is defined relatively in [alembic.ini](https://github.com/tulir/mautrix-telegram/blob/6ef31599e9690a31e24c059d6930979d6e943d43/alembic.ini#L5).

Reply:
It's unfortunately not documented in the bridge's wiki since this is usually not a problem when setting up the bridge in a git checkout of `mautrix-telegram`.

I simply did a `ln -sf ${pkgs.mautrix-telegram}/alembic alembic` with in a `preStart` hook of the `systemd` service back then. Unless you have any better ideas that should be the way to go :)

Reply:
I tried your suggestion Ma27 but I've this error :
`ln: failed to create symbolic link 'alembic': Read-only file system`

Reply:
Can you please share your local diff with me then?

Reply:
Sorry, I've just seen the last answer while reading unread GitHub notifications. In fact that's even better than my solution :+1: